bumble/hid.py

```diff
@@ -0,0 +1,295 @@
+        header = (HID_SET_PROTOCOL << 4 | self.protocol_mode)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.append(self.protocol_mode)
+        return bytes(packet_bytes)
+
+    def to_bytes_send_data(self) -> bytes:
+        header = ((HID_DATA << 4) | HID_OUTPUT_REPORT)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.extend(self.data)
+        return bytes(packet_bytes)
+# -----------------------------------------------------------------------------
+class HIDHost(EventEmitter):
+    l2cap_channel: Optional[l2cap.Channel]
+
+    def __init__(self, device: Device, connection: Connection) -> None:
+        super().__init__()
+        self.device = device
+        self.connection = connection
+        self.l2cap_ctrl_channel= None
+        self.l2cap_intr_channel = None
+
+        # Register ourselves with the L2CAP channel manager
+        device.register_l2cap_server(HID_CONTROL_PSM, self.on_connection)
+        device.register_l2cap_server(HID_INTERRUPT_PSM, self.on_connection)
+
+    async def connect_control_channel(self) -> None:
+        # Create a new L2CAP connection - control channel
+        try:
+            self.l2cap_ctrl_channel = await self.device.l2cap_channel_manager.connect(
+                self.connection, HID_CONTROL_PSM
+            )
+        except ProtocolError as error:
+            logger.error(f'L2CAP connection failed: {error}')
+            raise
+
+        assert self.l2cap_ctrl_channel is not None
+        # Become a sink for the L2CAP channel
+        self.l2cap_ctrl_channel.sink = self.on_ctrl_pdu
+
+    async def connect_interrupt_channel(self) -> None:
+        # Create a new L2CAP connection - interrupt channel
+        try:
+            self.l2cap_intr_channel = await self.device.l2cap_channel_manager.connect(
+                self.connection, HID_INTERRUPT_PSM
+            )
+        except ProtocolError as error:
+            logger.error(f'L2CAP connection failed: {error}')
+            raise
+
+        assert self.l2cap_intr_channel is not None
+        # Become a sink for the L2CAP channel
+        self.l2cap_intr_channel.sink = self.on_intr_pdu
+
+    async def disconnect_interrupt_channel(self) -> None:
+        if self.l2cap_intr_channel is None:
+            raise InvalidStateError('invalid state')
+        await self.l2cap_intr_channel.disconnect()  # type: ignore
+
+    async def disconnect_control_channel(self) -> None:
+        if self.l2cap_ctrl_channel is None:
+            raise InvalidStateError('invalid state')
+        await self.l2cap_ctrl_channel.disconnect()  # type: ignore
+
+    def on_connection(self, l2cap_channel: l2cap.Channel) -> None:
+        logger.debug(f'+++ New L2CAP connection: {l2cap_channel}')
+        l2cap_channel.on('open', lambda: self.on_l2cap_channel_open(l2cap_channel))
+
+    def on_l2cap_channel_open(self, l2cap_channel: l2cap.Channel) -> None:
```

Comment:
I don't see an opportunity for inlining here: when `on_connection` is invoked, the channel isn't open yet, it will be open after some transitions of the state machine, at which point `open` will be emitted. (unless I misunderstood the question)